src/core/include/mp-units/framework/customization_points.h

```diff
@@ -149,6 +149,42 @@ struct quantity_values {
   }
 };
 
+
+/**
+ * @brief A type trait that defines the behavior of scaling a value using a magnitude
+ *
+ * Whereas C++ numeric types usually represent a (fixed) subset of the real numbers
+ * (or another vector-space over the field of the real numbers),
+ * the magnitude concept fundamentally can represent any real number.
+ * Thus, in general, the result of a scaling operation is not exactly representable,
+ * and some form of approximation may be needed. That approximation is not
+ * part of the semantics of a physical quantitiy, but of its representation
+ * in C++. Therefore, the approximation semantics are dictatet by the
+ * representation type, which can be customised for user-types through
+ * this type-trait.
+ *
+ * In the following, $\mathcal{V}$ shall denote the vector-space represented by all representation
+ * types involved in the following discussion.
+ *
+ * A specialisation @c scaling_traits for a type @c Rep shall provide the following
+ * template static method members:
+ *  - `template <Magnitude M, typename From> static constexpr Rep scale_from(M scaling_factor, const From &value)`:
```

Comment:
Oh, that is a problem. The whole premise behind my refactoring here is that, fundamentally, magnitudes represent exact real numbers. That is, each of them is really a symbolic constant, and it should be up to the representation to find a good way to approximate the result of a multiplication with that symbolic constant. Thus, first approximating the scaling factor using a standard type in `get_value<U>` defeats the purpose. Non-standard types for `U` don't work, because providing an API for non-standard types to determine the return value `get_value<U>` is basically what we are discussing here.

In theory, we could define a "dynamic approximation interface", i.e. provide symbolic math for those symbols including a comparison operator, so that a multi-precision representation could choose to perhaps do a continued fraction expansion of that symbolic constant until it looses interest in more precision. In practice however it would be preferable to instead inspect the actual representation of the symbolic constant we already have.

In practice, all existing code paths do go through `get_value<U>` anyway. In particular, `fixed_point<std::int64_t>` promises to provide a full `i64 x i64.64 -> i64` rescaling operation which is able to attain full 64 bits of accuracy for scale factors close to one, but the factor is an approximation of the magnitude taken from `get_value<long double>`, so it will usually lead to 53 bits of accuracy (certainly on arm64).

Reply:
So, to summarise, the pragmatic approach goes through `long double`, and accepts that users of the public interface won't be able to achieve more than 53 bits of accuracy. Then, the minimal API surface to get there would be just `get_value<long double>(M)`. The current implementation also uses `is_integral(M)`, `is_integral(pow<-1>(M))` and `get_value<long double>(pow<-1>(M))`, but those three could be approximated using the return value of `get_value<long double>(M)` (to 53 bits accuracy ;-)).

In order to allow more precision, either for a custom representation but also for `std::int64_t`, currently, the blocker is access to more bits of precision for the irrationals, first and foremost, `pi`. With access to the guts of magnitude, I should be able to whip up a portable `constexpr` "wide float" implementation that can produce 64 bits of accuracy, and also single out any powers of pi for which to pull-in a 64 bit approximation as needed. But that's a task for the future.

Reply:
If the only thing that we need from magnitudes is `get_value<long double>(M)`, then we can just pass the resulting `long double` to scaling functions.

Reply:
It's good enough, as long as people are happy with the 53 bits of precision - it will therefore exclude extended precision representation types, e.g. from [mppp](https://github.com/bluescarni/mppp), or at least render those useless. But then, the design is sound and leaves a back-door for potential users of such a library to use magnitude internals to access more bits of precision of a conversion factor.